File: tests/naiveOracle.js
'use strict';

/**
 * Brute-force Voronoi oracle. Assigns every pixel of a width x height grid to
 * its nearest site under the supplied distance callback. This is
 * O(width * height * sites) and is kept only as a test reference; the shipped
 * library uses the Lee & Wong divide-and-conquer generator instead.
 */

/**
 * Generate a per-pixel color grid by brute force.
 *
 * @param {Array<Array<number>>} points - sites in the form [x,y]
 * @param {number} width
 * @param {number} height
 * @param {function} distanceCallback - (pointA, pointB) => number
 * @returns {Array<Array<number>>} one [r,g,b] per pixel
 */
function generateVoronoiPoints(points, width, height, distanceCallback){

    let colors = points.map(e =>{ return {point:e, color: new Array(3).fill(0).map(d => Math.ceil(Math.random() * 255))}})

    let imageData = new Array(width * height).fill(0).map((point, index) => {
        let coordinate = [index % height , Math.ceil(index / height)];
        let closest = colors.reduce((c,e) => {

            if(Array.isArray(c)){
                return c.every(d => distanceCallback(d.point, coordinate) < distanceCallback(e.point, coordinate) ) ? c : e;
            }
            else if(distanceCallback(c.point, coordinate) === distanceCallback(e.point, coordinate)){
                return [c,e];
            }
            else{
                return distanceCallback(c.point, coordinate) < distanceCallback(e.point, coordinate) ? c : e;
            }

        }, {point:[Infinity,Infinity]});

        return Array.isArray(closest) ? [0,0,0] : closest.color;
    });

    return imageData;
}

module.exports = {generateVoronoiPoints};

File: py_ai/naive_oracle.py
"""
Brute-force Voronoi oracle. Assigns every pixel of a width x height grid to
its nearest site under the supplied distance callback. This is
O(width * height * sites) and is kept only as a test reference; the ported
library uses the Lee & Wong divide-and-conquer generator instead.
"""

import math
import random


def generate_voronoi_points(points, width, height, distance_callback):
    """
    Generate Voronoi points via a basic, naive algorithm. Takes any distance callback

    :param points: list of points [[x,y], ...]
    :param width: integer width
    :param height: integer height
    :param distance_callback: function(point1, point2) returning distance
    :returns: list of color arrays
    """
    colors = [
        {"point": e, "color": [math.ceil(random.random() * 255) for _ in range(3)]}
        for e in points
    ]

    image_data = []
    for index in range(width * height):
        coordinate = [index % height, math.ceil(index / height)]
        closest = colors[0]

        for c in colors:
            if isinstance(closest, list):
                # This branch unlikely in Python port
                pass
            elif distance_callback(closest["point"], coordinate) == distance_callback(
                c["point"], coordinate
            ):
                closest = [closest, c]
            else:
                closest = (
                    c
                    if distance_callback(c["point"], coordinate)
                    < distance_callback(closest["point"], coordinate)
                    else closest
                )

        # Find actual closest point
        best = colors[0]
        for c in colors:
            if distance_callback(c["point"], coordinate) < distance_callback(
                best["point"], coordinate
            ):
                best = c

        image_data.append(
            best["color"] if isinstance(best, dict) and "color" in best else [0, 0, 0]
        )

    return image_data
